File: LeetCode/113路径总和Ⅱ.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) [][]int {
	ans := make([][]int, 0)
	calc(root, &ans, new([]int), targetSum)
	return ans
}

func calc(root *TreeNode, ans *[][]int, path *[]int, sum int) {
	if root == nil {
		return
	}
	sum -= root.Val
	*path = append(*path, root.Val)
	if root.Left == nil && root.Right == nil && sum == 0 {
		temp := make([]int, len(*path))
		for i := 0; i < len(*path); i++ {
			temp[i] = (*path)[i]
		}
		*ans = append(*ans, temp)
	}
	calc(root.Left, ans, path, sum)
	calc(root.Right, ans, path, sum)
	*path = (*path)[:len(*path)-1]
	return
}

/*
	Go传值，不是传引用
	new 使用来分配内存，并且返回内存地址，不必指定长度
	make 是创建channel map 和切片的内存分配，指定长度
*/

File: LeetCode/113路径总和Ⅱ.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> ans;
    List<Integer> path;

    public void calc(TreeNode root, int sum) {
        if (root == null) return ;
        path.add(root.val);
        sum -= root.val;
        if (root.left == null && root.right == null && sum == 0) {
            ans.add(new ArrayList<>(path));
        }
        if (root.left == null && root.right == null) return ;
        if (root.left != null) {
            calc(root.left, sum);
            path.remove(path.size() - 1);
        }
        if (root.right != null) {
            calc(root.right, sum);
            path.remove(path.size() - 1);
        }
        return ;
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        ans = new ArrayList<>();
        path = new ArrayList<>();
        if (root == null) return ans;
        calc(root, targetSum);
        return ans;
    }
}